subprojects/files/src/main/java/org/gradle/api/internal/file/archive/impl/StreamZipInput.java

```diff
@@ -42,31 +42,52 @@ public Iterator<ZipEntry> iterator() {
             protected ZipEntry computeNext() {
                 java.util.zip.ZipEntry nextEntry;
                 try {
-                    nextEntry = in.getNextEntry();
+                    nextEntry = inputStream.getNextEntry();
                 } catch (IOException e) {
                     throw new FileException(e);
                 }
-                return nextEntry == null ? endOfData() : new JdkZipEntry(nextEntry, new Supplier<InputStream>() {
-                    @Override
-                    public InputStream get() {
-                        return in;
-                    }
-                }, new Runnable() {
-                    @Override
-                    public void run()  {
-                        try {
-                            in.closeEntry();
-                        } catch (IOException e) {
-                            throw new FileException(e);
-                        }
-                    }
-                });
+                return nextEntry == null ? endOfData() : new StreamZipEntry(nextEntry);
             }
         };
     }
 
     @Override
     public void close() throws IOException {
-        in.close();
+        inputStream.close();
+    }
+
+    private class StreamZipEntry extends AbstractZipEntry {
+        private boolean opened;
+
+        public StreamZipEntry(java.util.zip.ZipEntry entry) {
+            super(entry);
+        }
+
+        @Override
+        public <T> T withInputStream(InputStreamAction<T> action) throws IOException {
+            if (opened) {
+                throw new IllegalStateException("The input stream for " + getName() + " has already been opened.  It cannot be reopened again.");
+            }
+
+            try {
+                return action.run(inputStream);
+            } finally {
+                opened = true;
```

Comment:
Nit: this should be set before the action is executed, no?

Reply:
Done.